go/pkg/tool/tool.go

```diff
@@ -31,54 +31,126 @@ const (
 	stderrFile = "stderr"
 )
 
+// testOnlyStartDeterminismExec
+var testOnlyStartDeterminismExec = func() {}
+
 // Client is a remote execution client.
 type Client struct {
 	GrpcClient *rc.Client
 }
 
+// CheckDeterminism executes the action the given number of times and compares
+// output digests, reporting failure if a mismatch is detected.
+func (c *Client) CheckDeterminism(ctx context.Context, actionDigest, inputRoot string, attempts int) error {
+	oe := outerr.SystemOutErr
+	fmc := filemetadata.NewNoopCache()
+	client := &rexec.Client{
+		FileMetadataCache: fmc,
+		GrpcClient:        c.GrpcClient,
+	}
+	cmd, err := c.prepCommand(ctx, client, actionDigest, inputRoot)
+	if err != nil {
+		return err
+	}
+	opt := &command.ExecutionOptions{AcceptCached: false, DownloadOutputs: false, DownloadOutErr: true}
+	firstRes, firstMd := client.Run(ctx, cmd, opt, oe)
+	for i := 1; i < attempts; i++ {
+		testOnlyStartDeterminismExec()
```

Comment:
Could we run them concurrently to save time?

Reply:
That's a good idea. Do you mind if I do this in a follow up PR? I'll file a bug for it.

Reply:
Sure!